readers-frontend/src/components/NavbarColourSwitcher.astro

```diff
@@ -9,56 +9,173 @@ interface Props {
 const { featuredCategories, style } = Astro.props;
 
 // A simplified category list for the navbar colour switcher script
-const scrollCategories = featuredCategories.map((category) => {
-  return {
-    categories_name: category.categories_name,
-    categories_backgroundColor: category.categories_backgroundColor,
-  };
-});
+const scrollCategories = featuredCategories.reduce((accumulator, category) => {
+  accumulator.set(
+    category.categories_name,
+    category.categories_backgroundColor || "#000"
+  );
+  return accumulator;
+}, new Map<string, string>());
+
+const serialisedScrollCategories = JSON.stringify(
+  Array.from(scrollCategories.entries())
+);
 ---
 
-<script define:vars={{ scrollCategories, style }}>
-  const desktopNav = document.querySelector("#desktopNav");
-  const mobileHeader = document.querySelector("#mobileHeader");
-  const mobileNouseLogo = document.querySelector("#mobileHeader img");
-  window.addEventListener("scroll", () => {
-    const offset = window.innerWidth >= 768 ? 60 : 120;
-    scrollCategories.forEach((category) => {
+<navbar-controller
+  data-categories={serialisedScrollCategories}
+  data-style={style}></navbar-controller>
+
+<script>
+  /**
+   * Custom element that controls Navbar colour changes
+   */
+  class NavbarController extends HTMLElement {
+    // Properties
+    scrollCategories: Map<string, string>;
+    assignedStyle: "muse" | "nouse" = "nouse";
+    desktopNav: HTMLElement | undefined;
+    mobileHeader: HTMLElement | undefined;
+    mobileNouseLogo: HTMLElement | undefined;
+    sections: NodeListOf<Element> | undefined;
+    observer: IntersectionObserver | undefined;
+
+    constructor() {
+      super();
+      // Get parameters
+      this.scrollCategories = new Map<string, string>(
```

Comment:
Just out of curiosity, is there a reason why we're using a custom HTML element? If we passed in the parameters to this script tag using `define:vars`, we won't have to use JSON stringify then JSON parse to pass the categories. This is a minor detail ofc but I'm interested to hear your thoughts.

Reply:
If you use `define:vars` the script will be inline. This is the same astro does with `<astro-island>` for JS framework components.

Reply:
That's a good point actually!